rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -111,33 +127,90 @@ rmw_create_node(const char * name)
     return NULL;
   }
 
-  rmw_node_t * node = new rmw_node_t;
+  rmw_node_t * node = rmw_node_allocate();
   node->implementation_identifier = rti_connext_dynamic_identifier;
   node->data = participant;
   return node;
 }
 
+rmw_ret_t
+rmw_destroy_node(rmw_node_t * node)
+{
+  if (!node) {
+    rmw_set_error_string("node handle is null");
+    return RMW_RET_ERROR;
+  }
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node handle,
+    node->implementation_identifier, rti_connext_dynamic_identifier,
+    return RMW_RET_ERROR)
+
+  DDSDomainParticipantFactory * dpf_ = DDSDomainParticipantFactory::get_instance();
+  if (!dpf_) {
+    rmw_set_error_string("failed to get participant factory");
+    return RMW_RET_ERROR;
+  }
+
+  DDSDomainParticipant * participant = static_cast<DDSDomainParticipant *>(node->data);
+  if (!participant) {
+    rmw_set_error_string("participant handle is null");
+  }
+  // This unregisters types and destroys topics which were shared between
+  // publishers and subscribers and could not be cleaned up in the delete functions.
+  participant->delete_contained_entities();
+
+  DDS_ReturnCode_t ret = dpf_->delete_participant(participant);
+  if (ret != DDS_RETCODE_OK) {
+    rmw_set_error_string("failed to delete participant");
+    return RMW_RET_ERROR;
+  }
+  rmw_node_free(node);
+  node = nullptr;
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+destroy_type_code(DDS_TypeCode * type_code)
+{
+  DDS_TypeCodeFactory * factory = NULL;
+  factory = DDS_TypeCodeFactory::get_instance();
+  if (!factory) {
+    rmw_set_error_string("failed to get typecode factory");
+    return RMW_RET_ERROR;
+  }
+
+  DDS_ExceptionCode_t ex;
+  factory->delete_tc(type_code, ex);
+  if (ex != DDS_NO_EXCEPTION_CODE) {
+    rmw_set_error_string("failed to delete type code struct");
+    return RMW_RET_ERROR;
+  }
+  // TODO(wjwwood): is this sufficient clean up of the type code?
```

Comment:
Is `destroy_type_code` invoked for every type code which is being created? If yes, I would assume that should be sufficient.

Or are there currently cases which it not invoked yet?